### ISSUE TITLE
Nodes unreachable under high memory; kubelet not evicting pods

### ISSUE THREAD

**Issue by asemarian** [k3s-io/k3s#11312]
Hello,

I’m running a Kubernetes cluster with k3s, and I’ve been experiencing intermittent issues with some of the nodes. Occasionally, nodes become unreachable, changing from “Ready” to “NotReady,” causing all workloads on them to be inaccessible. I often need to reboot the node to resolve the issue. The problem is very similar to the one described [here](https://www.reddit.com/r/kubernetes/comments/j7xw0e/node_oom_and_then_completely_unresponsive/).

In the most recent incident, I noticed that the Prometheus pod was consuming unusually high memory, which seemed to trigger the issue (no memory limits are set on the pod). My question is: why isn’t memory pressure kicking in and prompting the kubelet to evict pods? I’m using the [kube-hetzner](https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner) project, mostly with default settings. The node in question has 3 vCPUs and 4 GB RAM. These are the kubelet args taken from `/etc/rancher/k3s/config.yaml`:
```
"kubelet-arg":
- "cloud-provider=external"
- "volume-plugin-dir=/var/lib/kubelet/volumeplugins"
- "kube-reserved=cpu=50m,memory=300Mi,ephemeral-storage=1Gi"
- "system-reserved=cpu=250m,memory=300Mi"
```
It's worth mentioning that I couldn't reproduce the issue. I stress-tested the node but the pods are always killed before they make the node unstable.

Is there a configuration I can adjust to ensure the kubelet has enough headroom to start evicting pods before the node becomes completely unreachable? I’ve observed that, in another managed cluster we use, nodes never go down under similar conditions. Instead, the node is marked with `MemoryPressure` and the eviction process starts to prevent node instability.

Any insights on how to achieve similar resilience would be greatly appreciated! Thank you.

**Comment by brandond** [comment]
Does this node have swap enabled? Why aren't you setting memory limits that are at least lower than what's available on your node? Prometheus is pretty intense, I'm not sure I'd try to run it with less than 4gb allocated just to it, let alone running it on a node with only 4gb total.

**Comment by asemarian** [comment]
Hi @brandond. Thanks for the prompt response.

I don't want to give the impression that the problem is with Prometheus itself. To be clear, it does indeed use a lot of memory, somewhere in the range of 2.5Gi to be more precise. With that said, setting a memory limit did not resolve the issue. The pod was never OOM killed for whatever reason. What's more, the same issue occurred on other nodes with 8Gi of memory.

**Comment by asemarian** [comment]
I managed to reproduce the issue by cordoning all nodes except for one, then attempting to schedule as many pods as possible on that remaining node. It was really about the number of pods, not how much resources each one consumes.

I was able to resolve this issue by setting two kubelet args: I set `eviction-hard` to `imagefs.available<15%,memory.available<100Mi,nodefs.available<10%,pid.available<10%,nodefs.inodesFree<5%,imagefs.inodesFree<6%` and I also updated `kube-reserved` to `cpu=100m,memory=600Mi,ephemeral-storage=1Gi,pid=1024` effectively doubling the reserved CPU and memory in addition to reserving some PIDs. I'm still using the default value for `system-reserved`.
